flink-connector-cassandra/src/main/java/org/apache/flink/connector/cassandra/source/reader/CassandraSplitReader.java

```diff
@@ -0,0 +1,280 @@
+    }
+
+    @Override
+    public void handleSplitsChanges(SplitsChange<CassandraSplit> splitsChanges) {
+        for (CassandraSplit cassandraSplit : splitsChanges.splits()) {
+            unprocessedSplits.add(cassandraSplit.toSplitState());
+        }
+    }
+
+    @VisibleForTesting
+    static String getHighestSplitQuery(String query, String partitionKey, BigInteger highest) {
+        return generateQuery(
+                query, partitionKey, highest, " (token(%s) >= %d) AND", " WHERE (token(%s) >= %d)");
+    }
+
+    @VisibleForTesting
+    static String getLowestSplitQuery(String query, String partitionKey, BigInteger lowest) {
+        return generateQuery(
+                query, partitionKey, lowest, " (token(%s) < %d) AND", " WHERE (token(%s) < %d)");
+    }
+
+    @VisibleForTesting
+    static String generateRangeQuery(String query, String partitionKey) {
+        return generateQuery(
+                query,
+                partitionKey,
+                null,
+                " (token(%s) >= ?) AND (token(%s) < ?) AND",
+                " WHERE (token(%s) >= ?) AND (token(%s) < ?)");
+    }
+
+    private static String generateQuery(
+            String query,
+            String partitionKey,
+            @Nullable BigInteger token,
+            String whereFilter,
+            String noWhereFilter) {
+        Matcher queryMatcher = Pattern.compile(SELECT_REGEXP).matcher(query);
+        if (!queryMatcher.matches()) {
+            throw new IllegalStateException(
+                    String.format(
+                            "Failed to extract keyspace and table out of the provided query: %s",
+                            query));
+        }
+        final int whereIndex = query.toLowerCase().indexOf("where");
+        int insertionPoint;
+        String filter;
+        if (whereIndex != -1) {
+            insertionPoint = whereIndex + "where".length();
+            filter =
+                    (token == null)
+                            ? String.format(whereFilter, partitionKey, partitionKey)
+                            : String.format(whereFilter, partitionKey, token);
+        } else {
+            // end of keyspace.table
+            insertionPoint = queryMatcher.end(2);
+            filter =
+                    (token == null)
+                            ? String.format(noWhereFilter, partitionKey, partitionKey)
+                            : String.format(noWhereFilter, partitionKey, token);
+        }
+        return String.format(
+                "%s%s%s",
+                query.substring(0, insertionPoint), filter, query.substring(insertionPoint));
+    }
+
+    /**
+     * This method populates the {@code Map<String, Collection<CassandraRow>> recordsBySplit} map
+     * that is used to create the {@link RecordsBySplits} that are output by the fetch method. It
+     * modifies its {@code output} parameter.
+     */
+    private void addRecordsToOutput(
+            ResultSet resultSet,
```

Comment:
Why aren't we doing the row->pojo mapping here?

Reply:
Because in the javadoc of `SingleThreadMultiplexSourceReaderBase` the responsibilities are set as this:
`A RecordEmitter that takes a record from the Split Reader and updates the checkpointing state and converts it into the final form. For example for Kafka, the Record Emitter takes a ConsumerRecord, puts the offset information into state, transforms the records with the deserializers into the final type, and emits the record.` So  `CassandraSplitReader` deals with Cassandra `Row` and `CassandraRecordEmitter` deals with pojo.